src/RpcServer/Session.cs

```diff
@@ -41,6 +42,12 @@ public Session(NeoSystem system, byte[] script, Signer[] signers, Witness[] witn
                 Witnesses = witnesses
             };
             Engine = ApplicationEngine.Run(script, Snapshot, container: tx, settings: system.Settings, gas: gas, diagnostic: diagnostic);
+            if (Engine.State == VMState.HALT && tx != null)
+            {
+                tx.SystemFee = Engine.GasConsumed;
+                tx.NetworkFee = 1_00000000;
+                Engine = ApplicationEngine.Run(script, Snapshot, container: tx, settings: system.Settings, gas: Engine.GasConsumed, diagnostic: diagnostic);
```

Comment:
So, what if it fails?

Reply:
If it fails it will be a good change, because before it was `HALT`.
The flow looks like to be the same, @roman-khimov, nothing will change from the original flow, it just add a more precise gas limit

Reply:
> If it fails it will be a good change, because before it was HALT.

But what's good about it? Estimation failed? How do I send a transaction then?

I know these things are not likely to happen, but contracts checking fees shouldn't exist as well and yet they do.

> it just add a more precise gas limit

For normal contracts it changes nothing (except for additional execution overhead), for strange contracts that care about fee values it can lead to any result --- some will work, some will fail.